lib/spack/spack/spec.py

```diff
@@ -1518,7 +1518,7 @@ def root(self):
     @property
     def package(self):
         if not self._package:
-            self._package = spack.repo.get(self)
+            self._package = spack.repo.path.get(self)
```

Comment:
Is it possible to disallow calling `package` on a non-concrete spec?  I think this would be good for making the lifecycle clear, and for ensuring we don't proliferate package objects.

It seems like the places that use this in the tests (`buildrequest`, `buildtask`) just ensure that you can't call `install()` on an abstract spec.

Also, if we only call `get()` on concrete specs, we can key the `dict` by `dag_hash()` (or maybe we have to use `process_hash()`?) to avoid costly spec comparisons.

Reply:
Could always make concrete specs a subclass, actually strongly type it and only provide the method on a concrete spec object.  Perhaps not the most pythonic, but I tend to prefer to make guarantees like that true by construction rather than by unit test where feasible.  Would help with the virtual check problem too since a concrete instance could just override that with a shortcut.

Reply:
I think this is a good idea.

Reply:
I'm working on disallowing calling `Spec.package` with a non-concrete spec. I think that is doable, even though there are multiple changes to do. After that I'll try to make `spack.repo.get` do the same. That might be slightly more difficult due to some requirements to be able to work with abstract specs to compute the `content_hash`.

Reply:
Done in b89c1ef